src/m3/deserializer.ts

```diff
@@ -0,0 +1,200 @@
+import {
+    Concept,
+    ConceptInterface,
+    Containment,
+    Datatype,
+    Feature,
+    FeaturesContainer,
+    Link,
+    M3Concept,
+    Metamodel,
+    MetamodelElement,
+    PrimitiveType,
+    Property,
+    Reference
+} from "./types.ts"
+import {SingleRef} from "../references.ts"
+import {Id, Node} from "../types.ts"
+import {SerializedNode} from "../serialization.ts"
+import {metaConcepts, metaFeatures} from "./self-definition.ts"
+
+
+export const deserializeMetamodel = (serializedNodes: SerializedNode[]): Metamodel => {
+
+    const metamodelSerNode = serializedNodes.find(({type}) => type === metaConcepts.metamodel.id)
+    if (metamodelSerNode === undefined) {
+        throw new Error(`could not deserialize: no instance of LIonCore's Metamodel concept found in serialization`)
+    }
+
+    // make a map id -> serialized node:
+    const serializedNodeById: { [id: Id]: SerializedNode } = {}
+    serializedNodes.forEach((node) => {
+        serializedNodeById[node.id] = node
+    })
+
+    const deserializedNodeById: { [id: Id]: Node } = {}
+
+    const constructMemoised = (serNode: SerializedNode, parent?: M3Concept): M3Concept => {
+        const node = construct(serNode, parent)
+        deserializedNodeById[node.id] = node
+        return node
+    }
+
+    const mapConstructMemoised = <T extends Node>(parent: M3Concept) =>
+        (id: Id): T =>
+            constructMemoised(serializedNodeById[id], parent) as unknown as T
+
+    const referencesToInstall: [node: Node, featureName: string, refId: Id][] = []
+
+    const construct = ({type, id, properties, children, references}: SerializedNode, parent?: M3Concept): M3Concept => {
+        switch (type) {
+            case metaConcepts.concept.id: {
+                const {
+                    [metaFeatures.namespacedEntity_simpleName.id]: simpleName,
```

Comment:
I am unfamiliar with this TS construct. I am not sure if this is commonly known or it needs an explanation

Reply:
It's a JS construct even, and semantically equivalent to: `obj[<property name>] = <value>`.

Reply:
(I think it's common enough in {J|T}S world.)